apiserver/common/modelstatus.go

```diff
@@ -43,12 +44,13 @@ func ModelApplicationInfo(applications []Application) ([]params.ModelApplication
 }
 
 // NewModelStatusAPI creates an implementation providing the ModelStatus() API.
-func NewModelStatusAPI(backend ModelManagerBackend, machineService MachineService, authorizer facade.Authorizer, apiUser names.UserTag) *ModelStatusAPI {
+func NewModelStatusAPI(backend ModelManagerBackend, getMachineService func(coremodel.UUID) MachineService,
+	authorizer facade.Authorizer, apiUser names.UserTag) *ModelStatusAPI {
 	return &ModelStatusAPI{
-		authorizer:     authorizer,
-		apiUser:        apiUser,
-		backend:        backend,
-		machineService: machineService,
+		authorizer:        authorizer,
+		apiUser:           apiUser,
+		backend:           backend,
+		getMachineService: getMachineService,
```

Comment:
This is the correct approach. These "controller" access patterns should always pass in the model access "lazily". We should then attempt to remove "strict" access to model services either via state or the new domain services.

Attempting to have optimization paths for "if you're the same model as the facade" leads to woe and confusion. We should just have one way to do it. The way you've proposed here.